WeakAuras/Profiling.lua

```diff
@@ -360,3 +366,160 @@ function WeakAuras.PrintProfile()
+RealTimeProfilingWindow.bars = {}
+RealTimeProfilingWindow:SetMovable(true)
+
+local texture = "Interface\\DialogFrame\\UI-DialogBox-Background"
+function RealTimeProfilingWindow:AddBar(num, name, map, total)
+  local bar = self.bars[name]
+
+  -- hide bar if it's going to be out of background window
+  if self.titleHeight + self.barHeight * num > self.height then
+    if bar then
+      bar:Hide()
+    end
+    return
+  end
+
+  local pct = 100 * map.elapsed / total
+
+  if not bar then
+    bar = CreateFrame("FRAME", nil, self)
+    self.bars[name] = bar
+    Mixin(bar, SmoothStatusBarMixin)
+
+    local fg = bar:CreateTexture(nil, "ARTWORK")
+    fg:SetSnapToPixelGrid(false)
+    fg:SetTexelSnappingBias(0)
+    fg:SetTexture(texture)
+    fg:SetDrawLayer("ARTWORK", 0)
+    fg:Show()
+    bar.fg = fg
+
+    local bg = bar:CreateTexture(nil, "ARTWORK")
+    bg:SetSnapToPixelGrid(false)
+    bg:SetTexelSnappingBias(0)
+    bg:SetTexture(texture)
+    bg:SetDrawLayer("ARTWORK", -1)
+    bg:SetAllPoints()
+    bg:Show()
+    bar.bg = bg
+
+    bar:SetSize(self.width, self.barHeight)
+    if not bar.txtName then
+      local txtName = bar:CreateFontString(nil, "OVERLAY", "GameFontNormal")
+      bar.txtName = txtName
+      txtName:SetPoint("TOPLEFT", bar, "TOPLEFT")
+      txtName:SetPoint("BOTTOMRIGHT", bar, "BOTTOMRIGHT", -30)
+      txtName:SetJustifyH("LEFT")
+    end
+    if not bar.txtPct then
+      local txtPct = bar:CreateFontString(nil, "OVERLAY", "GameFontNormal")
+      bar.txtPct = txtPct
+      txtPct:SetPoint("TOPLEFT", bar, "TOPRIGHT", -50, 0)
+      txtPct:SetPoint("BOTTOMRIGHT", bar, "BOTTOMRIGHT")
+      txtPct:SetJustifyH("RIGHT")
+    end
+    bar:Show()
+  end
+  bar:ClearAllPoints()
+  bar:SetPoint("TOPLEFT", self, "TOPLEFT", 0, - (num - 1) * self.barHeight - self.titleHeight)
+  bar.fg:ClearAllPoints()
+  bar.fg:SetPoint("TOPLEFT", bar)
+  bar.fg:SetSize(self.width / 100 * pct, self.barHeight)
+  bar.txtName:SetText(("%s (%.2fms)"):format(name, map.elapsed))
+  bar.txtPct:SetText(("%.2f%%"):format(pct))
+  bar:Show()
+  return bar
+end
+
+function RealTimeProfilingWindow:RefreshBars()
+  if not profileData.systems.time then
+    return
+  end
+
+  local total = TotalProfileTime(WeakAuras.profileData.auras)
+  for i, k in ipairs(SortProfileMap(WeakAuras.profileData.auras)) do
+    if (k ~= "time" and k ~= "wa") then
+      local bar = self:AddBar(i, k, profileData.auras[k], total)
+      if bar then
+        if i % 2 == 0 then
+          bar.fg:SetColorTexture(1, 1, 1, 0.7)
+          bar.bg:SetColorTexture(0, 0, 0, 0.2)
+        else
+          bar.fg:SetColorTexture(1, 1, 1, 0.9)
+          bar.bg:SetColorTexture(0, 0, 0, 0.4)
+        end
+      end
+    end
+  end
+  local timespent = debugprofilestop() - profileData.systems.time.start
+  self.statsFrameText:SetText(("Time in WeakAuras: %.2f%%"):format(100 * profileData.systems.wa.elapsed / timespent))
+end
+
+function RealTimeProfilingWindow:Start()
```

Comment:
Here as well you have a lot of raw frame manip that distracts from the purpose of this method. It would be good imo to encapsulate that, with AceGUI widgets or with simple helper functions.

Reply:
Refactored both RefreshBars() and Start() methods